src/algorithms/tonal/pitchyinprobabilities.cpp

```diff
@@ -0,0 +1,333 @@
+  Real& RMS = _RMS.get();
+  
+  if (_precisetime) {
+    slowDifference(signal, _yin);
+  } else {
+    fastDifference(signal, _yin, size_t(_yin.size()));
+  }
+
+  // Compute a cumulative mean normalized difference function
+  _yin[0] = 1;
+  Real sum = 0.; 
+  for (int tau=1; tau < (int) _yin.size(); ++tau) {
+    sum += _yin[tau];
+    if (sum == 0) {
+      _yin[tau] = 1;
+    } else {
+      _yin[tau] *= tau / sum;
+    }
+
+    // Cannot simply check for sum==0 because NaN will be also produced by 
+    // infinitely small values 
+    // if (isnan(_yin[tau])) {
+    //   _yin[tau] = 1;
+    // }
+  }
+  // _yin[tau] is equal to 1 in the case if the df value for 
+  // this tau is the same as the mean across all df values from 1 to tau   
+
+  // Calculate YIN probabilities
+  static float betaDist2[100] = {0.012614,0.022715,0.030646,0.036712,0.041184,0.044301,0.046277,0.047298,0.047528,0.047110,0.046171,0.044817,0.043144,0.041231,0.039147,0.036950,0.034690,0.032406,0.030133,0.027898,0.025722,0.023624,0.021614,0.019704,0.017900,0.016205,0.014621,0.013148,0.011785,0.010530,0.009377,0.008324,0.007366,0.006497,0.005712,0.005005,0.004372,0.003806,0.003302,0.002855,0.002460,0.002112,0.001806,0.001539,0.001307,0.001105,0.000931,0.000781,0.000652,0.000542,0.000449,0.000370,0.000303,0.000247,0.000201,0.000162,0.000130,0.000104,0.000082,0.000065,0.000051,0.000039,0.000030,0.000023,0.000018,0.000013,0.000010,0.000007,0.000005,0.000004,0.000003,0.000002,0.000001,0.000001,0.000001,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000};
+
+  size_t minTau = 2;
+  size_t maxTau = _yin.size();
+
+  Real minWeight = 0.01;
+  size_t nThreshold = 100;
+  int nThresholdInt = nThreshold;
+
+  vector<Real> thresholds(nThresholdInt);
+  vector<Real> distribution(nThresholdInt);
+  _peakProb.assign(_yin.size(), 0.0);
+
+  for (int i = 0; i < nThresholdInt; ++i) {
+    distribution[i] = betaDist2[i];
+    thresholds[i] = 0.01 + i*0.01;
+  }
+
+  int currThreshInd = nThreshold-1;
+  size_t tau;
+  tau = minTau;
+  size_t minInd = 0;
+  Real minVal = 42.0;
+  Real sumProb = 0.0;
+  while (tau+1 < maxTau) {
+    if (_yin[tau] < thresholds[thresholds.size()-1] && _yin[tau+1] < _yin[tau]) {
+      while (tau + 1 < maxTau && _yin[tau+1] < _yin[tau]) {
+        tau++;
+      }
+
+      if (_yin[tau] < minVal && tau > 2) {
+        minVal = _yin[tau];
+        minInd = tau;
+      }
+
+      currThreshInd = nThresholdInt-1;
+      while (thresholds[currThreshInd] > _yin[tau] && currThreshInd > -1) {
+        _peakProb[tau] += distribution[currThreshInd];
+        currThreshInd--;
+      }
+      sumProb += _peakProb[tau];
+      tau++;
+    } else {
+      tau++;
+    }
+  }
+
+  if (_peakProb[minInd] > 1) {
+    E_WARNING("WARNING: yin has prob > 1 ??? I'm returning all zeros instead.");
```

Comment:
Improve this warning message

Reply:
This should never happen, right? If that's the case may be we should through an exception.